### PR TITLE
GDB-10418 Single Node Deployment

### PR DESCRIPTION
**Pull request by viktor-ribchev**
## Description

Added the ability to deploy a single node GraphDB

## Related Issues

GDB-10418 

## Changes

Added support for single node deployment. 
Removed hardcoded values for DNS zone in userdata scripts
Updated the NSGs ports based on graphdb_node_count
Updated user_data.tf
Changed the name of the LB target group to avoid conflicts when scaling from 1 to 3 AZs
Updated the monitoring, to not deploy cluster alarms when a single node is deployed
Updated the availability_tests
Updated how the VPC azs are calculated based on the graphdb_node_count
Added calculations for the subnets based on the graphdb_node_count

## Screenshots (if applicable)

N/A

## Checklist

- [X] I have tested these changes thoroughly.
- [X] My code follows the project's coding style.
- [X] I have added appropriate comments to my code, especially in complex areas.
- [X] All new and existing tests passed locally.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,13 @@
 # GraphDB AWS Terraform Module Changelog
 
+# 1.2.0
+* Added support for single node deployment
+  * Added new userdata script `10_start_graphdb_services.sh.tpl` for single node setup.
+  * Made cluster-related userdata scripts executable only when graphdb_node_count is greater than 1.
+* Removed hardcoded values from the userdata scripts.
+* Changed the availability tests http_string_type to be calculated based on TLS being enabled.
+
+
```
Comment (simeonzhekofff) [inline CHANGELOG.md:10]: Empty lines 😱

File: main.tf [hunk 2]
```diff
@@ -2,6 +2,17 @@ data "aws_region" "current" {}
 
 data "aws_caller_identity" "current" {}
 
+
```
Comment (simeonzhekofff) [inline main.tf:5]: Too much empty lines?

File: modules/graphdb/templates/00_functions.sh [hunk 3]
```diff
@@ -6,3 +6,127 @@
...
+    return 0
+  else
+    log_with_timestamp "GraphDB node $${gdb_address} is not available yet"
+    return 1
+  fi
+}
+
+
+
+
```
Comment (simeonzhekofff) [inline modules/graphdb/templates/00_functions.sh:47]: Too much empty lines

File: modules/graphdb/templates/10_start_single_graphdb_services.sh.tpl [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+set -o nounset
+set -o pipefail
+
+# Imports helper functions
+source /var/lib/cloud/instance/scripts/part-002
+
+NODE_DNS_RECORD=$(cat /var/opt/graphdb/node_dns)
+GRAPHDB_ADMIN_PASSWORD=$(aws --cli-connect-timeout 300 ssm get-parameter --region ${region} --name "/${name}/graphdb/admin_password" --with-decryption --query "Parameter.Value" --output text | base64 -d)
+RETRY_DELAY=5
+
```
Comment (simeonzhekofff) [inline modules/graphdb/templates/10_start_single_graphdb_services.sh.tpl:20]: Empty line

**Review by simeonzhekofff — APPROVED** [review]
LGTM

File: modules/graphdb/nsg.tf [hunk 5]
```diff
@@ -51,8 +54,8 @@ resource "aws_security_group_rule" "graphdb_network_lb_ingress" {
   description       = "CIRDs allowed to access GraphDB."
   security_group_id = aws_security_group.graphdb_security_group.id
   type              = "ingress"
-  from_port         = 7200
-  to_port           = 7200
+  from_port         = var.graphdb_node_count == 1 ? 7201 : 7200
```
Comment (mihailradkov) [inline modules/graphdb/nsg.tf:57]: Could calculate them once in some `local {}`

File: modules/load_balancer/main.tf [hunk 6]
```diff
@@ -1,6 +1,14 @@
+resource "random_id" "tg_name_suffix" {
```
Comment (mihailradkov) [inline modules/load_balancer/main.tf:1]: I see _what_ is going on but some comment for _why_ would be good.

File: main.tf [hunk 7]
```diff
@@ -2,6 +2,16 @@ data "aws_region" "current" {}
 
 data "aws_caller_identity" "current" {}
 
+locals {
+  # Reduce to one subnet if node_count is 1
+  effective_private_subnet_cidrs = var.graphdb_node_count == 1 ? [var.vpc_private_subnet_cidrs[0]] : var.vpc_private_subnet_cidrs
+  effective_public_subnet_cidrs  = var.graphdb_node_count == 1 ? [var.vpc_public_subnet_cidrs[0]] : var.vpc_public_subnet_cidrs
+  # Determine the appropriate subnets based on node_count
+  lb_subnets = var.graphdb_node_count == 1 ? (var.vpc_id == "" ? (var.lb_internal ? [module.vpc[0].private_subnet_ids[0]] : [module.vpc[0].public_subnet_ids[0]]) : (var.lb_internal ? [var.vpc_private_subnet_ids[0]] : [var.vpc_public_subnet_ids[0]])) : (var.vpc_id == "" ? (var.lb_internal ? module.vpc[0].private_subnet_ids : module.vpc[0].public_subnet_ids) : (var.lb_internal ? var.vpc_private_subnet_ids : var.vpc_public_subnet_ids))
```
Comment (mihailradkov) [inline main.tf:10]: Could format this a bit...

```hcl
  lb_subnets = var.graphdb_node_count == 1 ?
    (var.vpc_id == "" ? 
      (var.lb_internal ? [module.vpc[0].private_subnet_ids[0]] : [module.vpc[0].public_subnet_ids[0]]) : 
      (var.lb_internal ? [var.vpc_private_subnet_ids[0]] : [var.vpc_public_subnet_ids[0]])
    ) :
    (var.vpc_id == "" ? 
      (var.lb_internal ? module.vpc[0].private_subnet_ids : module.vpc[0].public_subnet_ids) : 
      (var.lb_internal ? var.vpc_private_subnet_ids : var.vpc_public_subnet_ids)
    )
```

Comment (viktor-ribchev) [reply main.tf]: done

File: modules/load_balancer/main.tf (hunk 6 above)
Comment (viktor-ribchev) [reply modules/load_balancer/main.tf:4]: done

File: modules/graphdb/nsg.tf (hunk 5 above)
Comment (viktor-ribchev) [reply modules/graphdb/nsg.tf]: done

**Review by mihailradkov — APPROVED** [review]

**Pull request merged**